javasource/restservices/util/JsonDeserializer.java

```diff
@@ -66,101 +66,180 @@ else if (jsonValue instanceof Long || jsonValue instanceof Double || jsonValue i
+		}
+
 		//string; autoresolve
 		else if (jsonValue instanceof String) {
 			if (!autoResolveReferences)
 				throw new RuntimeException("Unable to read url '" + jsonValue + "' into '" + targetType + "'; since references will not be resolved automatically for incoming data");
 			RestConsumer.request(context, HttpMethod.GET, (String) jsonValue, null, target, false);
 		}
-		
+
 		else if (jsonValue instanceof JSONObject) {
 			readJsonObjectIntoMendixObject(context, (JSONObject) jsonValue, target, autoResolveReferences);
 		}
-		
+
 		else
 			throw new RuntimeException("Unable to parse '" + jsonValue.toString() + "' into '" + targetType + "'");
 	}
-	
+
 	private static void readJsonObjectIntoMendixObject(IContext context, JSONObject object, IMendixObject target, boolean autoResolve) throws JSONException, Exception {
 		Iterator<String> it = object.keys();
 
 		Map<String, String> attributeNameMap = buildAttributeNameMap(target.getMetaObject());
-		
+
 		while(it.hasNext()) {
 			String attr = it.next();
 			String targetattr =  attributeNameMap.get(attr.toLowerCase().replaceAll("[^a-zA-Z0-9_]","_"));
-			
+
 			if (targetattr == null) {
 				if (RestServices.LOGUTIL.isDebugEnabled())
 					RestServices.LOGUTIL.debug("Skipping attribute '" + attr + "', not found in targettype: '" + target.getType() + "'");
 				continue;
 			}
-			
-			IMendixObjectMember<?> member = target.getMember(context, targetattr);
-			
-			if (member.isVirtual())
-				continue;
-			
-			//Reference
-			else if (member instanceof MendixObjectReference) {
-				String otherSideType = target.getMetaObject().getMetaAssociationParent(targetattr).getChild().getName();
-				if (!object.isNull(attr)) 
-					((MendixObjectReference)member).setValue(context, readJsonDataIntoMendixObject(context, object.get(attr), otherSideType, autoResolve));
-			}
-			
-			//ReferenceSet
-			else if (member instanceof MendixObjectReferenceSet){
-				String otherSideType = target.getMetaObject().getMetaAssociationParent(targetattr).getChild().getName();
-				JSONArray children = object.getJSONArray(attr);
-				List<IMendixIdentifier> ids = new ArrayList<IMendixIdentifier>();
-				
-				for(int i = 0; i < children.length(); i++) {
-					IMendixIdentifier child = readJsonDataIntoMendixObject(context, children.get(i), otherSideType, autoResolve);
-					if (child != null) {
-						/*
-						 * The core.createMendixIdentifier should be unnecessary, however, there is a bug there, see
-						 * support ticket 102188 
-						 */
-						ids.add(Core.createMendixIdentifier(child.toLong()));
+
+			if(target.hasMember(targetattr))
+			{
+				IMendixObjectMember<?> member = target.getMember(context, targetattr);
+
+				if (member.isVirtual())
+					continue;
+
+				//Reference
+				if (member instanceof MendixObjectReference) {
+					String otherSideType = target.getMetaObject().getMetaAssociationParent(targetattr).getChild().getName();
+					if(otherSideType.equals(target.getMetaObject().getName()))

```

Comment:
hmm what will happen if we are parsing a recursive data structure? I guess the otherSideType will still be the same type so it should be ok?